File: src/amedas_rainfall/indices/soil_tank.py
"""推定10分雨量と気象庁標準3段タンクモデルによる推定土壌雨量指数の計算（9節）。

重要な留意点（README/仕様書にも明記）:
    気象庁が公表する「土壌雨量指数」そのものではない。気象庁から取得可能なのは
    時別値のみであるため、1時間雨量を6等分して10分雨量を推定した上で、
    気象庁標準3段タンクモデルの係数体系を用いて独自に計算した値である。
    画面上・出力上は必ず「推定土壌雨量指数」と表示し、「気象庁公表土壌雨量指数」
    とは表示しない。

出典:
    気象庁「土壌雨量指数」に関する技術資料に示される標準3段タンクモデルの
    構造（3段直列タンク、各タンクの側方流出孔・側方流出係数・底面浸透係数）
    に基づく。具体的な数値は ``config/tank_model.yaml`` にまとめ、
    docs/calculation_method.md に出典・計算順序を記載する。
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
import pandas as pd

TEN_MIN_COLUMN = "rainfall_10min_mm"
TANK1_COLUMN = "soil_tank_1_mm"
TANK2_COLUMN = "soil_tank_2_mm"
TANK3_COLUMN = "soil_tank_3_mm"
SOIL_INDEX_COLUMN = "estimated_soil_rainfall_mm"


def disaggregate_hourly_to_10min(rainfall_used_mm: pd.Series) -> pd.Series:
    """閾値処理後時雨量を10分雨量へ均等分配する（1時間雨量を6等分）。

    気象庁の時別値は「当該時刻までの1時間」の積算値であるため、時刻 t の
    値は (t-50分, t-40分, ..., t) の6個の10分値に等分配する。
    時雨量が0（無降雨とみなされた場合を含む）であれば6個とも0とする。
    欠測（NaN）の時間は6個ともNaNとする。
    """
    parts = []
    for offset_min in (50, 40, 30, 20, 10, 0):
        shifted_index = rainfall_used_mm.index - pd.Timedelta(minutes=offset_min)
        part = pd.Series(rainfall_used_mm.to_numpy() / 6.0, index=shifted_index)
        parts.append(part)
    combined = pd.concat(parts).sort_index()
    combined.name = TEN_MIN_COLUMN
    return combined


@dataclass
class TankOutlet:
    height_mm: float
    coefficient_per_hour: float


@dataclass
class TankSpec:
    outlets: list[TankOutlet]
    infiltration_coefficient_per_hour: float


@dataclass
class TankModelConfig:
    time_step_hours: float
    tank1: TankSpec
    tank2: TankSpec
    tank3: TankSpec
    initial_storage_mm: dict[str, float] = field(
        default_factory=lambda: {"tank1": 0.0, "tank2": 0.0, "tank3": 0.0}
    )

    @classmethod
    def from_dict(cls, raw: dict) -> "TankModelConfig":
        def _spec(key: str) -> TankSpec:
            node = raw[key]
            outlets = [
                TankOutlet(height_mm=o["height_mm"], coefficient_per_hour=o["coefficient_per_hour"])
                for o in node["outlets"]
            ]
            return TankSpec(
                outlets=outlets,
                infiltration_coefficient_per_hour=node["infiltration_coefficient_per_hour"],
            )

        return cls(
            time_step_hours=raw.get("time_step_hours", 1.0 / 6.0),
            tank1=_spec("tank1"),
            tank2=_spec("tank2"),
            tank3=_spec("tank3"),
            initial_storage_mm=raw.get(
                "initial_storage_mm", {"tank1": 0.0, "tank2": 0.0, "tank3": 0.0}
            ),
        )


def _side_outflow_mm(storage_mm: float, spec: TankSpec, dt_hours: float) -> float:
    total = 0.0
    for outlet in spec.outlets:
        if storage_mm > outlet.height_mm:
            total += outlet.coefficient_per_hour * (storage_mm - outlet.height_mm) * dt_hours
    return total


def _infiltration_mm(storage_mm: float, spec: TankSpec, dt_hours: float) -> float:
    return spec.infiltration_coefficient_per_hour * storage_mm * dt_hours


def run_tank_model_10min(
    rainfall_10min_mm: pd.Series,
    config: TankModelConfig,
) -> pd.DataFrame:
    """10分刻みで3段タンクモデルを実行する。

    Returns:
        10分刻みのタンク貯留量・流出量・浸透量を持つDataFrame。
    """
    n = len(rainfall_10min_mm)
    dt_hours = config.time_step_hours
    index = rainfall_10min_mm.index
    values = rainfall_10min_mm.to_numpy(dtype=float)

    s1 = np.full(n, np.nan)
    s2 = np.full(n, np.nan)
    s3 = np.full(n, np.nan)
    outflow1 = np.full(n, np.nan)
    outflow2 = np.full(n, np.nan)
    outflow3 = np.full(n, np.nan)
    infil1 = np.full(n, np.nan)
    infil2 = np.full(n, np.nan)
    infil3 = np.full(n, np.nan)

    cur1 = config.initial_storage_mm.get("tank1", 0.0)
    cur2 = config.initial_storage_mm.get("tank2", 0.0)
    cur3 = config.initial_storage_mm.get("tank3", 0.0)
    pending_reset = True

    for i in range(n):
        rain = values[i]
        if np.isnan(rain):
            pending_reset = True
            continue
        if pending_reset:
            cur1 = config.initial_storage_mm.get("tank1", 0.0)
            cur2 = config.initial_storage_mm.get("tank2", 0.0)
            cur3 = config.initial_storage_mm.get("tank3", 0.0)
            pending_reset = False

        # タンク1: 入力は降雨
        storage1 = cur1 + rain
        out1 = _side_outflow_mm(storage1, config.tank1, dt_hours)
        inf1 = _infiltration_mm(storage1, config.tank1, dt_hours)
        cur1 = max(storage1 - out1 - inf1, 0.0)

        # タンク2: 入力はタンク1の浸透
        storage2 = cur2 + inf1
        out2 = _side_outflow_mm(storage2, config.tank2, dt_hours)
        inf2 = _infiltration_mm(storage2, config.tank2, dt_hours)
        cur2 = max(storage2 - out2 - inf2, 0.0)

        # タンク3: 入力はタンク2の浸透
        storage3 = cur3 + inf2
        out3 = _side_outflow_mm(storage3, config.tank3, dt_hours)
        inf3 = _infiltration_mm(storage3, config.tank3, dt_hours)
        cur3 = max(storage3 - out3 - inf3, 0.0)

        s1[i] = cur1
        s2[i] = cur2
        s3[i] = cur3
        outflow1[i] = out1
        outflow2[i] = out2
        outflow3[i] = out3
        infil1[i] = inf1
        infil2[i] = inf2
        infil3[i] = inf3

    return pd.DataFrame(
        {
            TANK1_COLUMN: s1,
            TANK2_COLUMN: s2,
            TANK3_COLUMN: s3,
            "tank1_outflow_mm": outflow1,
            "tank2_outflow_mm": outflow2,
            "tank3_outflow_mm": outflow3,
            "tank1_infiltration_mm": infil1,
            "tank2_infiltration_mm": infil2,
            "tank3_infiltration_mm": infil3,
        },
        index=index,
    )


def aggregate_tank_result_to_hourly(tank_10min: pd.DataFrame) -> pd.DataFrame:
    """10分刻みのタンク計算結果を、各時間の6回目更新後の値として時別化する。"""
    hourly = tank_10min.resample("h", label="right", closed="right").last()
    hourly[SOIL_INDEX_COLUMN] = (
        hourly[TANK1_COLUMN] + hourly[TANK2_COLUMN] + hourly[TANK3_COLUMN]
    )
    return hourly[[TANK1_COLUMN, TANK2_COLUMN, TANK3_COLUMN, SOIL_INDEX_COLUMN]]


def calculate_estimated_soil_rainfall_index(
    rainfall_used_mm: pd.Series,
    config: TankModelConfig,
) -> tuple[pd.DataFrame, pd.DataFrame]:
    """推定10分雨量からタンクモデルを実行し、10分値と時別値の両方を返す。

    Returns:
        (10分刻みDataFrame, 時別DataFrame) のタプル。
    """
    rainfall_10min = disaggregate_hourly_to_10min(rainfall_used_mm)
    tank_10min = run_tank_model_10min(rainfall_10min, config)
    tank_hourly = aggregate_tank_result_to_hourly(tank_10min)
    return tank_10min, tank_hourly

File: core/src/soil_tank.rs
//! 推定10分雨量と気象庁標準3段タンクモデルによる推定土壌雨量指数の計算
//! （docs/calculation_method.md 6節・7節）。
//!
//! 移行の主目的: Python版（純Pythonのforループ、数十年分で約52秒）を
//! コンパイル言語に置き換えて高速化する、性能上の最重要モジュール。

use serde::Deserialize;

#[derive(Debug, Clone, Deserialize)]
pub struct TankOutlet {
    pub height_mm: f64,
    pub coefficient_per_hour: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct TankSpec {
    pub outlets: Vec<TankOutlet>,
    pub infiltration_coefficient_per_hour: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct InitialStorageMm {
    pub tank1: f64,
    pub tank2: f64,
    pub tank3: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct TankModelConfig {
    pub time_step_hours: f64,
    pub tank1: TankSpec,
    pub tank2: TankSpec,
    pub tank3: TankSpec,
    pub initial_storage_mm: InitialStorageMm,
}

pub struct TankModelResult {
    pub tank1_mm: Vec<f64>,
    pub tank2_mm: Vec<f64>,
    pub tank3_mm: Vec<f64>,
    pub tank1_outflow_mm: Vec<f64>,
    pub tank2_outflow_mm: Vec<f64>,
    pub tank3_outflow_mm: Vec<f64>,
    pub tank1_infiltration_mm: Vec<f64>,
    pub tank2_infiltration_mm: Vec<f64>,
    pub tank3_infiltration_mm: Vec<f64>,
}

/// 閾値処理後時雨量を10分雨量へ均等分配する（1時間雨量を6等分）。
/// 欠測(NaN)の時間は6個ともNaNとする。
pub fn disaggregate_hourly_to_10min(hourly_mm: &[f64]) -> Vec<f64> {
    let mut out = Vec::with_capacity(hourly_mm.len() * 6);
    for &v in hourly_mm {
        let part = if v.is_nan() { f64::NAN } else { v / 6.0 };
        for _ in 0..6 {
            out.push(part);
        }
    }
    out
}

fn side_outflow_mm(storage_mm: f64, spec: &TankSpec, dt_hours: f64) -> f64 {
    spec.outlets
        .iter()
        .map(|o| {
            if storage_mm > o.height_mm {
                o.coefficient_per_hour * (storage_mm - o.height_mm) * dt_hours
            } else {
                0.0
            }
        })
        .sum()
}

fn infiltration_mm(storage_mm: f64, spec: &TankSpec, dt_hours: f64) -> f64 {
    spec.infiltration_coefficient_per_hour * storage_mm * dt_hours
}

/// 10分刻みで3段タンクモデルを実行する。
///
/// Python版 (`indices/soil_tank.py::run_tank_model_10min`) と1対1で対応する。
/// 各タンクについて、(1)入力を加算→(2)側方流出量・浸透量を計算→
/// (3)貯留量から両方を差し引く（0未満にならないようクリップ）、の順で計算する。
/// 欠測(NaN)の直後の最初の有効値で、3タンクとも初期貯留量から再初期化する。
pub fn run_tank_model_10min(rainfall_10min_mm: &[f64], config: &TankModelConfig) -> TankModelResult {
    let n = rainfall_10min_mm.len();
    let dt = config.time_step_hours;

    let mut s1 = vec![f64::NAN; n];
    let mut s2 = vec![f64::NAN; n];
    let mut s3 = vec![f64::NAN; n];
    let mut out1v = vec![f64::NAN; n];
    let mut out2v = vec![f64::NAN; n];
    let mut out3v = vec![f64::NAN; n];
    let mut infil1v = vec![f64::NAN; n];
    let mut infil2v = vec![f64::NAN; n];
    let mut infil3v = vec![f64::NAN; n];

    let mut cur1 = config.initial_storage_mm.tank1;
    let mut cur2 = config.initial_storage_mm.tank2;
    let mut cur3 = config.initial_storage_mm.tank3;
    let mut pending_reset = true;

    for i in 0..n {
        let rain = rainfall_10min_mm[i];
        if rain.is_nan() {
            pending_reset = true;
            continue;
        }
        if pending_reset {
            cur1 = config.initial_storage_mm.tank1;
            cur2 = config.initial_storage_mm.tank2;
            cur3 = config.initial_storage_mm.tank3;
            pending_reset = false;
        }

        let storage1 = cur1 + rain;
        let out1 = side_outflow_mm(storage1, &config.tank1, dt);
        let inf1 = infiltration_mm(storage1, &config.tank1, dt);
        cur1 = (storage1 - out1 - inf1).max(0.0);

        let storage2 = cur2 + inf1;
        let out2 = side_outflow_mm(storage2, &config.tank2, dt);
        let inf2 = infiltration_mm(storage2, &config.tank2, dt);
        cur2 = (storage2 - out2 - inf2).max(0.0);

        let storage3 = cur3 + inf2;
        let out3 = side_outflow_mm(storage3, &config.tank3, dt);
        let inf3 = infiltration_mm(storage3, &config.tank3, dt);
        cur3 = (storage3 - out3 - inf3).max(0.0);

        s1[i] = cur1;
        s2[i] = cur2;
        s3[i] = cur3;
        out1v[i] = out1;
        out2v[i] = out2;
        out3v[i] = out3;
        infil1v[i] = inf1;
        infil2v[i] = inf2;
        infil3v[i] = inf3;
    }

    TankModelResult {
        tank1_mm: s1,
        tank2_mm: s2,
        tank3_mm: s3,
        tank1_outflow_mm: out1v,
        tank2_outflow_mm: out2v,
        tank3_outflow_mm: out3v,
        tank1_infiltration_mm: infil1v,
        tank2_infiltration_mm: infil2v,
        tank3_infiltration_mm: infil3v,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_config() -> TankModelConfig {
        TankModelConfig {
            time_step_hours: 1.0 / 6.0,
            tank1: TankSpec {
                outlets: vec![TankOutlet { height_mm: 15.0, coefficient_per_hour: 0.1 }],
                infiltration_coefficient_per_hour: 0.12,
            },
            tank2: TankSpec {
                outlets: vec![TankOutlet { height_mm: 15.0, coefficient_per_hour: 0.05 }],
                infiltration_coefficient_per_hour: 0.05,
            },
            tank3: TankSpec {
                outlets: vec![TankOutlet { height_mm: 15.0, coefficient_per_hour: 0.01 }],
                infiltration_coefficient_per_hour: 0.01,
            },
            initial_storage_mm: InitialStorageMm { tank1: 0.0, tank2: 0.0, tank3: 0.0 },
        }
    }

    #[test]
    fn known_hand_calculation_single_step() {
        let config = test_config();
        let result = run_tank_model_10min(&[10.0], &config);
        let dt = 1.0 / 6.0;
        let expected_infil1 = 0.12 * 10.0 * dt;
        let expected_tank1 = 10.0 - expected_infil1;
        assert!((result.tank1_mm[0] - expected_tank1).abs() < 1e-12);
    }

    #[test]
    fn storage_never_negative() {
        let config = test_config();
        let values = vec![0.0, 1.0, 5.0, 20.0, 0.0, 0.0];
        let result = run_tank_model_10min(&values, &config);
        assert!(result.tank1_mm.iter().all(|&v| v >= 0.0));
        assert!(result.tank2_mm.iter().all(|&v| v >= 0.0));
        assert!(result.tank3_mm.iter().all(|&v| v >= 0.0));
    }
}
